### PR TITLE
Add sprite and mesh alteration examples

### LINKED ISSUE #3130: Examples to demonstrate changing a sprite and texture
## How can Bevy's documentation be improved?

Newcomers are often a bit lost when trying to learn how to idiomatically work with assets and handles.

The most common stumbling block is "how do I change a sprite once loaded?".

We should add two sibling examples, one for sprites, and one for meshes:

- show how to query for sprite / mesh handles, and change what they point to
- show how to programatically change the properties of a sprite / mesh (e.g. color, reflectance, translucency or so on)
- show how to mutate the underlying asset to e.g. add more vertexes

### PR DESCRIPTION
**Pull request by richchurcher**
# Objective

Add examples for manipulating sprites and meshes by either mutating the handle or direct manipulation of the asset, as described in #15056.

Closes #3130.

(The previous PR suffered a Git-tastrophe, and was unceremoniously closed, sry! :sweat_smile: )

### CODE REVIEW

File: examples/asset/alter_mesh.rs [hunk 1]
```diff
@@ -0,0 +1,200 @@
...
+
+        // Obtain a mutable reference to the Mesh asset.
+        let Some(mesh) = meshes.get_mut(handle) else {
+            return;
+        };
+
+        // Now we can directly manipulate vertices on the mesh. Here, we're just scaling in and out
+        // for demonstration purposes. This will affect all entities currently using the asset.
+        if let Some(VertexAttributeValues::Float32x3(positions)) =
+            mesh.attribute_mut(Mesh::ATTRIBUTE_POSITION)
```
Comment (richchurcher) [inline examples/asset/alter_mesh.rs:183]: I will freely admit this took quite a lot of getting to a "working" state, so if there's a more sensible way to do it let me know! Much head-scratching involved :laughing:

Comment (alice-i-cecile) [reply examples/asset/alter_mesh.rs:183]: Yep, our mesh modification API is not particularly ergonomic: it's a very rare operation to do in code!

**Review by alice-i-cecile — APPROVED** [review]
Perfect: this demonstrates the point really nicely.

File: examples/asset/alter_mesh.rs [hunk 2]
```diff
@@ -0,0 +1,200 @@
...
+                "Return: mutate the mesh itself, changing all copies of it",
+                TextStyle::default(),
+            ));
+        });
+}
+
+#[derive(Default)]
+struct IsMeshScaled(bool);
+
+fn keyboard_controls(
```
Comment (janhohenheim) [inline examples/asset/alter_mesh.rs:138]: Could you split this into two systems so that they can have meaningful names?
Doing so also allows you to completely skip the `keyboard_input` param by using a `input_just_pressed` run condition :)

File: examples/asset/alter_mesh.rs [hunk 3]
```diff
@@ -0,0 +1,200 @@
...
+    mut materials: ResMut<Assets<StandardMaterial>>,
+) {
+    let left_shape = Shape::Cube;
+    let right_shape = Shape::Cube;
+
+    let left_shape_model = asset_server.load(
+        GltfAssetLabel::Primitive {
+            mesh: 0,
+            primitive: 0,
+        }
```
Comment (janhohenheim) [inline examples/asset/alter_mesh.rs:51]: Maybe a very quick explanation as to what a "primitive" is in this context would be good

File: examples/asset/alter_mesh.rs [hunk 4]
```diff
@@ -0,0 +1,200 @@
...
+    commands.spawn((
+        Name::new("Camera"),
+        Camera3dBundle {
+            transform: Transform::from_xyz(0.0, 3.0, 20.0).looking_at(Vec3::ZERO, Vec3::Y),
+            ..default()
+        },
+    ));
+}
+
+fn instructions(mut commands: Commands) {
```
Comment (janhohenheim) [inline examples/asset/alter_mesh.rs:108]: This is usually called `spawn_text` or similar in other examples. Can we keep them consistent?

File: examples/asset/alter_mesh.rs (hunk 1 above)
Comment (janhohenheim) [reply examples/asset/alter_mesh.rs:183]: Pretty sure this this is the only way to do this, yeah

File: examples/asset/alter_mesh.rs [hunk 5]
```diff
@@ -0,0 +1,200 @@
...
+            ));
+        });
+}
+
+#[derive(Default)]
+struct IsMeshScaled(bool);
+
+fn keyboard_controls(
+    asset_server: Res<AssetServer>,
+    mut is_mesh_scaled: Local<IsMeshScaled>,
```
Comment (janhohenheim) [inline examples/asset/alter_mesh.rs:140]: This would be simpler if we just use a `Local<bool>`, imo

File: examples/asset/alter_mesh.rs (hunk 1 above)
Comment (janhohenheim) [inline examples/asset/alter_mesh.rs:183]: Maybe a comment about what `Float32x3` and `ATTRIBUTE_POSITION` mean would be nice?

**Review by janhohenheim — CHANGES_REQUESTED** [review]
Good idea! The `alter_mesh` example could mayyyybe be turned into `procedural_mesh`, as that is what most people using this API will probably end up doing.
The rest of the comments are just about how to make the examples a bit easier to understand for outsiders. Comments are mostly valid for both examples.
Good job with the examples, they're pretty nicely written!

File: examples/asset/alter_mesh.rs (hunk 1 above)
Comment (janhohenheim) [reply examples/asset/alter_mesh.rs:183]: Pretty sure this this is the only way to do this, yeah

File: examples/asset/alter_mesh.rs (hunk 5 above)
Comment (richchurcher) [reply examples/asset/alter_mesh.rs:140]: Oh, yeah :facepalm:

File: examples/asset/alter_mesh.rs (hunk 2 above)
Comment (richchurcher) [reply examples/asset/alter_mesh.rs:138]: Wait, that's a thing? Nice!

File: examples/asset/alter_mesh.rs (hunk 4 above)
Comment (richchurcher) [reply examples/asset/alter_mesh.rs:108]: Oh, TIL. Can do!

File: examples/asset/alter_mesh.rs (hunk 1 above)
Comment (richchurcher) [reply examples/asset/alter_mesh.rs:183]: Right, that'll mean _understanding_ what they mean :laughing: OK, on it.

Comment (janhohenheim) [reply examples/asset/alter_mesh.rs:183]: Let me know if you need a primer, I can weite something uo tomorrow if you want :)

Comment (richchurcher) [reply examples/asset/alter_mesh.rs:183]: No it's better for me if I figure it out!

**Comment by JMS55** [comment]
I'm surprised not to see any usage of RenderAssetUsages here.

**Comment by richchurcher** [comment]
> I'm surprised not to see any usage of RenderAssetUsages here.

I'm sorry, you might need to ELI5 here... how should I best use this?

**Comment by richchurcher** [comment]
> could mayyyybe be turned into procedural_mesh, as that is what most people using this API will probably end up doing. 

Couldn't make a decision either way, so left it alone.

**Comment by JMS55** [comment]
> I'm sorry, you might need to ELI5 here... how should I best use this?

Read the docs https://dev-docs.bevyengine.org/bevy/render/render_asset/struct.RenderAssetUsages.html, and then let me know if you have any questions.

I just realized the default is actually RenderAssetUsages::all() so that's why this example is fine, but I'd prefer if we could explicitly set that to all() and call out why it's necessary, as I _very_ frequently see people make the mistake of using RENDER_WORLD only and then being confused why their asset does not exist.

File: examples/asset/alter_mesh.rs [hunk 6]
```diff
@@ -0,0 +1,228 @@
...
+        }
+        .from_asset(left_shape.get_model_path()),
+        // This is the default loader setting, ensuring both `RENDER_WORLD` and `MAIN_WORLD` are
+        // enabled. It's provided explicitly here only by way of demonstration.
+        //
+        // A common mistake is to use `RENDER_WORLD` by itself, which can cause a confusing lack of
+        // assets available in `Res<Assets<Mesh>>`. `RENDER_WORLD` alone will cause the asset to be
+        // unloaded from the asset server after it's been sent to the GPU. Unless you have a clear
+        // reason not to do so, it's best to use `RenderAssetUsages::all().
+        |settings: &mut GltfLoaderSettings| settings.load_meshes = RenderAssetUsages::all(),
```
Comment (richchurcher) [inline examples/asset/alter_mesh.rs:75]: @JMS55 Is this the kind of thing you had in mind? New territory for me, so feel free to provide more direction!

**Review by janhohenheim — APPROVED** [review]
Looks great now, thanks :) The example are very readable, good job!

File: examples/asset/alter_mesh.rs (hunk 6 above)
Comment (janhohenheim) [reply examples/asset/alter_mesh.rs:75]: @JMS55 will know better than me, but IME it is the other way around: usually you won't be mutating meshes at runtime, so it's best to leave it at RENDER_WORLD only unless you have a good reason to set it to ALL.

**Comment by janhohenheim** [comment]
Leaving it as waiting for author because of the render world thing

File: examples/asset/alter_mesh.rs (hunk 6 above)
Comment (richchurcher) [reply examples/asset/alter_mesh.rs:75]: So JMS55 wrote:
> but I'd prefer if we could explicitly set that to all() and call out why it's necessary, as I very frequently see people make the mistake of using RENDER_WORLD only and then being confused why their asset does not exist.
Perhaps if I make a point of only setting this to `all` for the asset whose mesh we're going to directly modify, and re-word the comment to say the default is "usually safe" unless you have more specific needs? I'll push that so you can have a read.

Comment (janhohenheim) [reply examples/asset/alter_mesh.rs:78]: The default is *always* safe, as the default is ALL. Many people choose to use RENDER_WORLD instead of ALL for meshes, which is safe until you need to access the mesh on the CPU, which we do in this example.

File: examples/asset/alter_mesh.rs [hunk 7]
```diff
@@ -0,0 +1,231 @@
...
+        }
+        .from_asset(left_shape.get_model_path()),
+        // This is the default loader setting, ensuring both `RENDER_WORLD` and `MAIN_WORLD` are
+        // enabled. It's provided explicitly here only by way of demonstration. Unless you have
+        // specific requirements, it's usually safe to use the default.
+        //
+        // When needing to access a `Mesh` asset via `Res<Assets<Mesh>>`, a common mistake is to
+        // use `RENDER_WORLD` by itself, which can cause the asset to be missing from the resource.
+        // (`RENDER_WORLD` without `MAIN_WORLD` will result in the asset being unloaded from the
+        // asset server after it's been sent to the GPU.)
```
Comment (janhohenheim) [inline examples/asset/alter_mesh.rs:77]:
```suggestion
        // `RenderAssetUsages::all()` is already the default, so this line could be omitted.
        // We leave it in for explicitness however. The `RenderAssetUsages` tell Bevy whether to
        // keep the data around for the GPU, which corresponds to `RenderAssetUsages::RENDER_WORLD`,
        // or for the CPU, which corresponds to `RenderAssetUsages::MAIN_WORLD`. 
        // `RENDER_WORLD` is necessary to render the mesh, `MAIN_WORLD` is necessary to inspect
        // and modify the mesh. Since most games will not need to modify meshes at runtime, many users
        // opt to pass only `RENDER_WORLD`. That is more memory efficient, as we don't need to keep the mesh
        // in the RAM. For this example however, this would not work, as we need to inspect and modify the mesh
        // via `Res<Assets<Mesh>>` at runtime. That's the entire point of the example, after all!
```

File: examples/asset/alter_mesh.rs (hunk 6 above)
Comment (richchurcher) [reply examples/asset/alter_mesh.rs:78]: I'm with you now. OK, so maybe the second one we explicitly set it to `RENDER_WORLD` and add a further comment there. Maybe also modify the sprite example.

**Comment by richchurcher** [comment]
Thanks for being patient, I think I've arrived at a good place with the comments now!

**Comment by alice-i-cecile** [comment]
I'm going to merge this once CI is green :)

**Comment by richchurcher** [comment]
Argh internal imports again! One moment, caller.

**Comment by richchurcher** [comment]
@alice-i-cecile sorted

**Comment by janhohenheim** [comment]
> Thanks for being patient

If this took a moment for you, that's a big indicator that our existing docs could be better in this regard. Happy to have an example showing and explaining the concept now :)
Fantastic work!
